DOC-2078 (#219)
tigergraph/gsql-docs: modules/querying/pages/query-operations.adoc

If the user has set a working graph, then `DROP ALL` removes all the queries for that graph.
If a user has set their scope to be global, then `DROP ALL` removes all queries across all graph spaces.

== Query Descriptions

> Query descriptions explains the procedures for updating, showing, and dropping descriptions of queries and query parameters.

Reviewer: ```suggestion
Query descriptions explain the procedures for updating, showing, and dropping descriptions of queries and query parameters.
```

=== Update description
After creating a query, you can add a description for the query.

*Syntax*
[source]
----
UPDATE DESCRIPTION OF <objectType> <objectName> [ON GRAPH <graphName>] <description>
----

Required privilege
`WRITE_QUERY`

To update a description, run the `UPDATE DESCRIPTION` command as shown below. You can update the description of a query or a query parameter. If you choose not to specify a graph in the command, the current scope will be used as the scope of the description.

[source, gsql]
----

> GSQL > USE GRAPH POC_GRAPH

Reviewer: In other places in our documentation, "POC_GRAPH" is lower case: "poc_graph".
If this is just the name of an example graph, it's clearer if it's not all capital letters, because we want the reader to see the difference between what is a keyword and what is not.

GSQL > UPDATE DESCRIPTION OF QUERY query1 "xxx"
GSQL > UPDATE DESCRIPTION OF QUERY query2 ON GRAPH graph1 "xxx"

> GSQL > UPDATE DESCRIPTION OF QUERY_PARAM query1.param1 "xxx"

Reviewer: If these are supposed to be examples, why not put more realistic examples instead of "xxx"?
In particular, each of these 3 should have a different description.

----

[NOTE]
====
The term `UPDATE` here means adding a description to the target.
If the target does not already have a description, it will be added.
If the target already has a description, it will be replaced with the new one.
====

[CAUTION]
====
Using create or replace is preferable for updating a query to ensure that the description is retained.

> Otherwise, there is a risk of losing the description.

Reviewer: I was a briefly confused by this, because the actual meaning here is `CREATE OR REPLACE QUERY`, not just the regular English works "create or replace".  Please write it as code. (Maybe you were just copying what the developer wrote and didn't realize this is code?)

Also, I think it's clearer to a reader if the sentence starts by established the context:

When updating a query, it is recommended to use `CREATE OR REPLACE QUERY` instead of simply `CREATE QUERY` to ensure the description is retained.

By the way, this warning worries/confused me. Is the developer saying the product is unreliable?

Author: @victorleeTG 
This is what engineer has suggested:

The context is that when we use `CREATE QUERY` in GraphStudio, if the query already exist, it will first drop the query(which will drop the query's description together) then recreate the query. So this will losing all the descriptions about the query. It's better to use `CREATE OR REPLACE QUERY` then it will not drop the query.

====

=== Show description

*Syntax*
[source]
----
SHOW DESCRIPTION OF <objectType> <objectName> [ON GRAPH <graphName>]
----

Required privilege
`READ_QUERY`

To display a description, use the `SHOW DESCRIPTION` command as shown below. Similar to the update command, you can display the description of a query or a query parameter.

[source, gsql]
----
GSQL > USE GRAPH POC_GRAPH
GSQL > SHOW DESCRIPTION OF QUERY query1

> GSQL > SHOW DESCRIPTION OF QUERY query1,query2

Reviewer: This example shows that is is possible to have a comma-separated list of queries or parameters. The syntax above does not indicate that, so technically, the syntax description is wrong, or at least overly simplified.

A simply way to do this would be to replace <objectName> with <objectName(s)>

and below the syntax, add
"where `<objectName(s)>` can be a comma-separated list of names.

We can do this sort of natural language commentary with this "informal" syntax. We could not do it with EBNF, which has very strict rules. ENBF is supposed to be understood by a computer program, e.g. a syntax checker could read your EBNF rules and your code, and then tell you if you code follows the EBNF rules.

GSQL > SHOW DESCRIPTION OF QUERY * ON GRAPH graph1
GSQL > SHOW DESCRIPTION OF QUERY_PARAM query1.param1
GSQL > SHOW DESCRIPTION OF QUERY_PARAM query1.param1,query1.param2
GSQL > SHOW DESCRIPTION OF QUERY_PARAM query2.*
----

Support is provided for displaying descriptions of multiple targets simultaneously. You can separate the targets using commas, or use the wildcard `*` to display all descriptions.

=== Drop description

*Syntax*
[source]
----
DROP DESCRIPTION OF <objectType> <objectName> [ON GRAPH <graphName>]
----

Required privilege
`WRITE_QUERY`

To drop a description, run the `DROP DESCRIPTION` command like below.
Similar to the update command, you can drop the description of a global type object or a local type object.

To remove a description, use the DROP DESCRIPTION command as shown below.

> Similar to the update command, you can remove the description of a query or a query parameter.

Reviewer: This seems repetitive/redundant.  Is the second paragraph trying to say something new that the first paragraph didn't already say?

Tip: In regular English, the base meaning of "drop" is to fall, but in databases it has a particular technical interpretation.  Whereas "remove" means to take away. We want to use very straightforward, non-metaphorical English in our explanations. Therefore, I would use "remove" to explain what `DROP` means.

[source, gsql]
----
GSQL > USE GRAPH POC_GRAPH
GSQL > DROP DESCRIPTION OF QUERY query1
GSQL > DROP DESCRIPTION OF QUERY query1*,query2*
GSQL > DROP DESCRIPTION OF QUERY * ON GRAPH poc_graph
GSQL > DROP DESCRIPTION OF QUERY_PARAM query1.param1
GSQL > DROP DESCRIPTION OF QUERY_PARAM query1.param1,query1.param2

> GSQL > DROP DESCRIPTION OF QUERY_PARAM query1*

Reviewer: This example has `query1*` whereas the one above has `query2.*`:
one include a dot before the `*`, and one does not.  Are both supported, or is this is mistake?

Overall, I had a lot more comments on this than I thought I would. Are you read this as carefully as though you are trying to learn this well enough so that you could confidently write correct code?

To update a description, run the `UPDATE DESCRIPTION` command as shown below. You can update the description of a query or a query parameter. If you choose not to specify a graph in the command, the current scope will be used as the scope of the description.

[source, gsql]
----
GSQL > USE GRAPH poc_graph
GSQL > UPDATE DESCRIPTION OF QUERY query1 "List of all employees in the company"
GSQL > UPDATE DESCRIPTION OF QUERY query2 ON GRAPH graph1 "Details of employees in a specific department"
GSQL > UPDATE DESCRIPTION OF QUERY_PARAM query1.param1 "Department where the employee works"
----

[NOTE]
====
The term `UPDATE` here means adding a description to the target.
If the target does not already have a description, it will be added.
If the target already has a description, it will be replaced with the new one.
====

[CAUTION]
====
* When updating a query, use `CREATE OR REPLACE QUERY` instead of `CREATE QUERY` to retain the query description.
* `CREATE QUERY` may unintentionally drop both the query and its description if it already exists, whereas `CREATE OR REPLACE QUERY` ensures the description is retained by replacing the query without dropping it.
====

=== Show description

> The `SHOW DESCRIPTION` command displays the description of a query or a query parameter.

Reviewer: ```suggestion
The `SHOW DESCRIPTION` command displays the descriptions of the queries or query parameters that match the given `objectName`.
```

*Syntax*
[source]
----
SHOW DESCRIPTION OF <objectType> <objectName> [ON GRAPH <graphName>]
----

Required privilege
`READ_QUERY`

To display a description, use the `SHOW DESCRIPTION` command as shown below. Similar to the update command, you can display the description of a query or a query parameter.

[source, gsql]
----
GSQL > USE GRAPH poc_graph
GSQL > SHOW DESCRIPTION OF QUERY query1
GSQL > SHOW DESCRIPTION OF QUERY query1*
GSQL > SHOW DESCRIPTION OF QUERY * ON GRAPH poc_graph
GSQL > SHOW DESCRIPTION OF QUERY_PARAM query1.param1
GSQL > SHOW DESCRIPTION OF QUERY_PARAM query2.*
----

> * For the `SHOW DESCRIPTION` command, you can display the descriptions of individual queries or query parameters. Each query or parameter must be specified separately, and you cannot combine them in a single request.

Reviewer: ```suggestion
```
We already defined SHOW DESCRIPTION above.  
There is no need to say "must be specified separately" if we clearly define what you CAN do in one statement. Focus on ABILITIES, not INABILITIES.  Only mention inabilities if the feature gap is surprising.

> * The wildcard `*` can be used to show descriptions of all queries or all parameters for a specific query. 

Reviewer: ```suggestion
* The wildcard `*` can be used to match multiple queries or parameters.
** `*` by itself means all queries in this graph.
**`prefix*` means all queries whose name starts with `prefix`.
** `queryName.*` means all parameters of `queryName`.
```

> * `SHOW DESCRIPTION` also supports a wider range of regular expressions, allowing for more advanced querying when searching for descriptions.

Reviewer: ```suggestion
```
Again, this was an answer to a specific question that we asked Adil. It will make no sense to a reader:
"A wider range of regular expressions" --> wider than what? You didn't say.  Why is that even important to know?

=== Drop description

> The `DROP DESCRIPTION` command is used to remove descriptions of queries or query parameters. 

Reviewer: ```suggestion
The `DROP DESCRIPTION` removes descriptions of queries or query parameters. 
```

Required privilege
`WRITE_QUERY`

To remove a description, run the `DROP DESCRIPTION` command as shown below. 

[source, gsql]
----
GSQL > USE GRAPH poc_graph
GSQL > DROP DESCRIPTION OF QUERY query1
GSQL > DROP DESCRIPTION OF QUERY * ON GRAPH poc_graph
GSQL > DROP DESCRIPTION OF QUERY_PARAM query1.param1
GSQL > DROP DESCRIPTION OF QUERY_PARAM query1.param1,query1.param2
GSQL > DROP DESCRIPTION OF QUERY_PARAM query1.*
----

Support is provided for dropping descriptions of multiple targets simultaneously.
You can use the wildcard `*` to remove descriptions of all queries or all parameters of a specific query.
The query parameter's description is independent of the query's description.
Therefore, after dropping the query description, the query parameter's description remains unaffected.

[NOTE]
====
* `SHOW DESCRIPTION` allows you to display descriptions of multiple queries or query parameters using wildcards like `*` (to show all queries) or `prefix*` (to show queries starting with a specific prefix). You can also use `queryName.*` to show all parameters of a specific query.

> * `DROP DESCRIPTION` supports specifying multiple queries or query parameters in a single command using a comma-separated list. However, it only supports the wildcard `*` for selecting all queries or query parameters and does not support wildcards based on prefixes or individual query names.

Reviewer: Please reorganize to **focus** on the differences. I would make the bullets about the features, not about the commands.  E.g.
1. Wildcards:  SHOW ... , but DROP...
2. List of objects:  SHOW ..., but DROP ...

If the user has set a working graph, then `DROP ALL` removes all the queries for that graph.
If a user has set their scope to be global, then `DROP ALL` removes all queries across all graph spaces.

== Query Descriptions

> Query descriptions explain the procedures for updating, showing, and dropping descriptions of queries and query parameters.

Reviewer: I think you mixed up two separate things into one sentence:
1) What is a query description?
2) What is contained in this section of the documentation?
It would nice to have one concise sentence for each of these.

* `CREATE QUERY` may unintentionally drop both the query and its description if it already exists, whereas `CREATE OR REPLACE QUERY` ensures the description is retained by replacing the query without dropping it.
====

=== Show description

*Syntax*
[source]
----
SHOW DESCRIPTION OF <objectType> <objectName(s)> [ON GRAPH <graphName>]
----

`<objectName(s)>` can include a comma-separated list of names.

Required privilege
`READ_QUERY`

To display a description, use the `SHOW DESCRIPTION` command as shown below. Similar to the update command, you can display the description of a query or a query parameter.

[source, gsql]
----
GSQL > USE GRAPH poc_graph
GSQL > SHOW DESCRIPTION OF QUERY query1
GSQL > SHOW DESCRIPTION OF QUERY query1,query2
GSQL > SHOW DESCRIPTION OF QUERY * ON GRAPH graph1
GSQL > SHOW DESCRIPTION OF QUERY_PARAM query1.param1
GSQL > SHOW DESCRIPTION OF QUERY_PARAM query1.param1,query1.param2
GSQL > SHOW DESCRIPTION OF QUERY_PARAM query2.*
----

Support is provided for displaying descriptions of multiple targets simultaneously. You can separate the targets using commas, or use the wildcard `*` to display all descriptions.

=== Drop description

*Syntax*
[source]
----

> DROP DESCRIPTION OF <objectType> <objectName> [ON GRAPH <graphName>]

Reviewer: ```suggestion
DROP DESCRIPTION OF <objectType> <objectName(s)> [ON GRAPH <graphName>]
```

----

Required privilege
`WRITE_QUERY`

To remove a description, run the `DROP DESCRIPTION` command as shown below. This command can be used to remove the description of both global and local type objects, as well as queries and query parameters.

[source, gsql]
----
GSQL > USE GRAPH poc_graph
GSQL > DROP DESCRIPTION OF QUERY query1

> GSQL > DROP DESCRIPTION OF QUERY query1*,query2*

Reviewer: `GSQL > DROP DESCRIPTION OF QUERY query1*,query2*`

I think we have another question for a developer. 
What does `query1*` mean?  It seems to say "a name that begins with `query1` and then has any other characters. For example, `query1A`, `query1_is_the_best` and `query1_is_my_favorite` would fit this pattern.

In all the other examples, `*` could only be used to represent an entire name, not part of a name. My guess is that either `*` actually cannot be used for part of a name OR it ALWAYS can be used for the suffix of a name. We need the developer to clarify

GSQL > DROP DESCRIPTION OF QUERY * ON GRAPH poc_graph
GSQL > DROP DESCRIPTION OF QUERY_PARAM query1.param1

> GSQL > DROP DESCRIPTION OF QUERY_PARAM query1.param1,query1.param2

Reviewer: While we are talking to the developer, in  SHOW DESCRIPTION QUERY_PARAM and DROP DESCRIPTION QUERY_PARAM,
if there is a list of multiple parameters, can they come from different queries, e.g.
`query1.param1, query2.param2`?
